x-pack/plugins/fleet/server/services/security/message_signing_service.ts

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { generateKeyPairSync, createSign, randomBytes } from 'crypto';
+
+import type { KibanaRequest } from '@kbn/core-http-server';
+import type { SavedObjectsClientContract } from '@kbn/core-saved-objects-api-server';
+import type { EncryptedSavedObjectsClient } from '@kbn/encrypted-saved-objects-plugin/server';
+import { SECURITY_EXTENSION_ID } from '@kbn/core-saved-objects-server';
+
+import { MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE } from '../../constants';
+import { appContextService } from '../app_context';
+
+interface MessageSigningKeys {
+  private_key: string;
+  public_key: string;
+  passphrase: string;
+}
+
+export interface MessageSigningServiceInterface {
+  generateKeyPair(providedPassphrase?: string): Promise<void>;
+  sign(serializedMessage: Buffer | object): Promise<{ data: Buffer; signature: string }>;
+  getPublicKey(): Promise<string>;
+}
+
+export class MessageSigningService implements MessageSigningServiceInterface {
+  #esoClient: EncryptedSavedObjectsClient;
+  #_soClient: SavedObjectsClientContract | undefined;
+
+  constructor(esoClient: EncryptedSavedObjectsClient) {
+    this.#esoClient = esoClient;
+  }
+
+  public async generateKeyPair(providedPassphrase?: string) {
+    const currentKeyPair = await this.#getCurrentKeyPair();
+    if (currentKeyPair.privateKey && currentKeyPair.publicKey && currentKeyPair.passphrase) {
+      return Promise.resolve();
```

Comment:
(nit) An empty `return` is equivalent to `return Promise.resolve()` in an `async` function, so adding the `Promise.resolve` seems a little needless. Think we can simply `return;` here to bail out.